File: packages/translation/ios/Plugin/TranslationPlugin.swift
import Foundation
import Capacitor
import MLKitTranslate

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(TranslationPlugin)
public class TranslationPlugin: CAPPlugin {
    public let errorLanguageMissing = "language must be provided."
    public let errorTextMissing = "text must be provided."
    public let errorSourceLanguageMissing = "sourceLanguage must be provided."
    public let errorTargetLanguageMissing = "targetLanguage must be provided."
    public let errorDownloadModelFailed = "downloadModel failed."

    private var implementation: Translation?

    override public func load() {
        implementation = Translation(plugin: self)

        NotificationCenter.default.addObserver(
            forName: .mlkitModelDownloadDidSucceed,
            object: nil,
            queue: nil
        ) { [weak self] notification in
            guard let strongSelf = self else {
                return
            }
            strongSelf.handleModelDownloadDidSucceed(notification: notification)
        }

        NotificationCenter.default.addObserver(
            forName: .mlkitModelDownloadDidFail,
            object: nil,
            queue: nil
        ) { [weak self] notification in
            guard let strongSelf = self else {
                return
            }
            strongSelf.handleModelDownloadDidFail(notification: notification)
        }
    }

    @objc func deleteDownloadedModel(_ call: CAPPluginCall) {
        guard let languageOption = call.getString("language") else {
            call.reject(errorLanguageMissing)
            return
        }
        let language = TranslateLanguage(rawValue: languageOption)

        implementation?.deleteDownloadedModel(language: language, completion: { error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        })
    }

    @objc func downloadModel(_ call: CAPPluginCall) {
        guard let languageOption = call.getString("language") else {
            call.reject(errorLanguageMissing)
            return
        }
        let language = TranslateLanguage(rawValue: languageOption)

        implementation?.downloadModel(language: language, completion: { error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        })
    }

    @objc func getDownloadedModels(_ call: CAPPluginCall) {
        let models = implementation?.getDownloadedModels() ?? []
        var languagesResult = JSArray()
        for model in models {
            languagesResult.append(model.language.rawValue)
        }
        call.resolve([
            "languages": languagesResult
        ])
    }

    @objc func translate(_ call: CAPPluginCall) {
        guard let text = call.getString("text") else {
            call.reject(errorTextMissing)
            return
        }
        guard let sourceLanguageOption = call.getString("sourceLanguage") else {
            call.reject(errorSourceLanguageMissing)
            return
        }
        let sourceLanguage = TranslateLanguage(rawValue: sourceLanguageOption)
        guard let targetLanguageOption = call.getString("targetLanguage") else {
            call.reject(errorTargetLanguageMissing)
            return
        }
        let targetLanguage = TranslateLanguage(rawValue: targetLanguageOption)

        implementation?.translate(text: text, sourceLanguage: sourceLanguage, targetLanguage: targetLanguage, completion: { translatedText, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            call.resolve([
                "text": translatedText ?? ""
            ])
        })
    }

    @objc private func handleModelDownloadDidSucceed(notification: Notification) {
        guard let userInfo = notification.userInfo,
              let model = userInfo[ModelDownloadUserInfoKey.remoteModel.rawValue]
                as? TranslateRemoteModel else {
            return
        }
        implementation?.handleModelDownloadDidSucceed(model: model)
    }

    @objc private func handleModelDownloadDidFail(notification: Notification) {
        guard let userInfo = notification.userInfo,
              let model = userInfo[ModelDownloadUserInfoKey.remoteModel.rawValue]
                as? TranslateRemoteModel else {
            return
        }
        implementation?.handleModelDownloadDidFail(model: model)
    }
}

File: packages/translation/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/translation/TranslationPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.translation;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.nl.translate.TranslateRemoteModel;
import java.util.Set;

@CapacitorPlugin(name = "Translation")
public class TranslationPlugin extends Plugin {

    public static final String ERROR_LANGUAGE_MISSING = "language must be provided.";
    public static final String ERROR_TEXT_MISSING = "text must be provided.";
    public static final String ERROR_SOURCE_LANGUAGE_MISSING = "sourceLanguage must be provided.";
    public static final String ERROR_TARGET_LANGUAGE_MISSING = "targetLanguage must be provided.";

    private Translation implementation = new Translation();

    @PluginMethod
    public void deleteDownloadedModel(PluginCall call) {
        try {
            String language = call.getString("language");
            if (language == null) {
                call.reject(ERROR_LANGUAGE_MISSING);
                return;
            }

            implementation.deleteDownloadedModel(
                language,
                new DeleteDownloadedModelResultCallback() {
                    @Override
                    public void success() {
                        call.resolve();
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(message, exception);
            call.reject(message);
        }
    }

    @PluginMethod
    public void downloadModel(PluginCall call) {
        try {
            String language = call.getString("language");
            if (language == null) {
                call.reject(ERROR_LANGUAGE_MISSING);
                return;
            }

            implementation.downloadModel(
                language,
                new DownloadModelResultCallback() {
                    @Override
                    public void success() {
                        call.resolve();
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(message, exception);
            call.reject(message);
        }
    }

    @PluginMethod
    public void getDownloadedModels(PluginCall call) {
        try {
            implementation.getDownloadedModels(
                new GetDownloadedModelsResultCallback() {
                    @Override
                    public void success(Set<TranslateRemoteModel> models) {
                        JSArray languagesResult = new JSArray();
                        for (TranslateRemoteModel model : models) {
                            languagesResult.put(model.getLanguage());
                        }
                        JSObject result = new JSObject();
                        result.put("languages", languagesResult);
                        call.resolve(result);
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(message, exception);
            call.reject(message);
        }
    }

    @PluginMethod
    public void translate(PluginCall call) {
        try {
            String text = call.getString("text");
            if (text == null) {
                call.reject(ERROR_TEXT_MISSING);
                return;
            }
            String sourceLanguage = call.getString("sourceLanguage");
            if (sourceLanguage == null) {
                call.reject(ERROR_SOURCE_LANGUAGE_MISSING);
                return;
            }
            String targetLanguage = call.getString("targetLanguage");
            if (targetLanguage == null) {
                call.reject(ERROR_TARGET_LANGUAGE_MISSING);
                return;
            }

            implementation.translate(
                text,
                sourceLanguage,
                targetLanguage,
                new TranslateResultCallback() {
                    @Override
                    public void success(String text) {
                        JSObject result = new JSObject();
                        result.put("text", text);
                        call.resolve(result);
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(message, exception);
            call.reject(message);
        }
    }
}
